### PR TITLE
Fix - Express payments that have errors cause the whole block to stop working

### LINKED ISSUE #42256: Express payments that have errors cause the whole block to stop working
## Describe the bug

When an express payment method registers, it must define an `edit` component, this will display in the site editor, and a `content` component which will display on the front end.

If these components error then the express payments block will break too, in both contexts.

image

## To reproduce

1. Install and activate [WooCommerce PayPal Payments](https://wordpress.org/plugins/woocommerce-paypal-payments/) - set it up and ensure it's working on your site, ensure the express payment button renders in the editor and the front end. You will need to see this to verify the issue and the fix.
2. Clone the development version of [WooCommerce Stripe Payment Gateway](https://github.com/woocommerce/woocommerce-gateway-stripe) 
3. Build the plugin, following the instructions in the repo, then activate and enable it. You will also need to add your stripe keys to ensure this works. 
4. Run `npm run start` in the plugin directory.
5. Open the plugin in your editor, specifically at this file: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8ffd22aff3b06eda02a1ae2fd8368b71450b36a9/client/blocks/payment-request/index.js#L11
6. Redefine `ApplePayPreview` like so:
```ts
const ApplePayPreview = () => {
  const x = window.null.something.broken;
  return <img src={ applePayImage } alt="" disabled={x} />;
}
```
7. Open the site editor, go to Templates > Cart and then Templates > Checkout
8. Notice the block is broken in both templates. Notice the PayPal button is not displayed.
9. Undo your edits to https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8ffd22aff3b06eda02a1ae2fd8368b71450b36a9/client/blocks/payment-request/index.js#L11
10. Go to https://github.com/woocommerce/woocommerce-gateway-stripe/blob/4f47751f5dd3d26eb245878a936f96510a11d2e4/client/blocks/payment-request/payment-request-express.js#L95C1-L95C1 and add:
```ts
  const y = window.null.broken;
  console.log( y );
```
11. Open the Cart and Checkout block on the front end and notice the error, notice neither button appears. Try as both a logged-in user and a logged-out user.

## Expected behavior

The express payments block should render the remaining methods even if one breaks.

### PR DESCRIPTION
**Pull request by tarunvijwani**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #42256.

With this PR, we're enhancing

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate [WooCommerce PayPal Payments](https://wordpress.org/plugins/woocommerce-paypal-payments/) - set it up and ensure it's working on your site, ensure the express payment button renders in the editor and the front end. 
2. Clone the development version of [WooCommerce Stripe Payment Gateway](https://github.com/woocommerce/woocommerce-gateway-stripe) 
3. Build the plugin, following the instructions in the repo, then activate and enable it. 
4. Run `npm run start` in the stripe plugin directory.
5. Ensure Stripe's express payment button renders in the editor and the front end.
6. Open the plugin in your editor, specifically at this file: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8ffd22aff3b06eda02a1ae2fd8368b71450b36a9/client/blocks/payment-request/index.js#L11
7. Redefine `ApplePayPreview` like so:
```ts
const ApplePayPreview = () => {
const x = window.null.something.broken;
return <img src={ applePayImage } alt="" disabled={x} />;
}
```
8. Go to https://github.com/woocommerce/woocommerce-gateway-stripe/blob/4f47751f5dd3d26eb245878a936f96510a11d2e4/client/blocks/payment-request/payment-request-express.js#L95C1-L95C1 and add:
```ts
  const y = window.null.broken;
  console.log( y );
```
9. Open the site editor, go to Templates > Checkout
10. Confirm the error notice and PayPal button is visible in the express checkout block. 
11. Open the site editor and go to Templates > Cart.
9. Confirm the error notice and PayPal button is visible in the express checkout block. 
10. Visit the site as admin.
12. Add a product to the Cart and visit the Cart Block page.
13. Confirm the error notice and PayPal button is visible in the express checkout block. 
14. Navigate to the Checkout block page.
15. Confirm the error notice and PayPal button is visible in the express checkout block. 
16. Visit your site as a Guest. 
17. Add a product to the Cart and visit the Cart Block page.
18. Confirm only the PayPal button is visible in the express checkout block. 
19. Navigate to the Checkout block page.
20. Confirm only the PayPal button is visible in the express checkout block. 

https://github.com/woocommerce/woocommerce/assets/11503784/f0f7c83a-8f85-44c8-ba51-5a0ee033163c

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 
Improve handling of express checkout: Now, error messages for specific express payment methods won't affect the functionality of others.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/packages/components/store-notices-container/index.tsx
Comment (tarunvijwani) [inline plugins/woocommerce-blocks/packages/components/store-notices-container/index.tsx:104]: This is to make sure it displays the error notices. 
For express checkout notices, the notice type was neither default nor snackbar, so notices were not getting displayed. 

@opr This is why error notices of express checkout were not visible during our pair programming session.

**Review by opr — CHANGES_REQUESTED** [review]
@tarunvijwani Thanks for this PR.

For anyone picking up this work while Tarun is away, I think wrapping it in an error boundary is a great idea, however I would like to rethink what this error boundary displays. For admin users it shows the error which is great:

image

but for shoppers it shows a fairly useless message:

image

The next steps for this PR would be to remove that message from the shopper-facing part of the error boundary and instead show nothing.

I would also recommend hiding the entire express payment area if there is only one method there and it errors, but this may need to be looked at in another PR to avoid expanding the scope of this.

**Comment by ralucaStan** [comment]
Thank you @opr for the thorough review 👏 

Merchant view: I agree the notice is good for merchants but there is some room for improvement:
- we should specify which Payment is crashing, we have that data; the message should tell the merchant to contact those plugin maintainers
- the notice should not be displayed in the button listing. but above it. The result is not ideal. I see that the ErrorBoundry is at `<li>` level so the PR will need some logic to move that error a level up
- this extra logic would enable us to also not display the Express Payments container in case no payment button works, which by the way should be handled in this PR as well as we need to cover all cases

Shopper view:
The shopper message is not ideal because this would prompt them to contact us, without actually giving them a contact link.

I agree that the shopper should not see this notice, nor the Express Payments area if no button is working.

**Review by ralucaStan — CHANGES_REQUESTED** [review]
I've https://github.com/woocommerce/woocommerce/pull/42988#issuecomment-1971498535 for the approach taken by this PR. Design should also 👍 the final decision. Marking the PR as a draft.

File: plugins/woocommerce-blocks/packages/components/store-notices-container/index.tsx
Comment (nerrad) [reply plugins/woocommerce-blocks/packages/components/store-notices-container/index.tsx:104]: I would suggest trying to understand why the notice.type is not set since it looks like the [type should be `default` by default](https://github.com/WordPress/gutenberg/blob/a0cb09b99624d30ffff26dee7160292ab505655d/packages/notices/src/store/actions.js#L69-L81) when a notice is created. I'd also want to know _what_ the type is for these missing notices that your fixing. Perhaps it's being intentionally set to something different than "default" and that's why it's not being displayed because the current implementation only considers whether a notice is one of two types. There are two potential paths as a result of that investigation:

- Explicitly expect and handle whether a notice is a snackbar type or not. In which case, you can simplify things by changing the filter callback condition in the original code.
- Discover that there is a specific type being passed in for a created notice that isn't handled yet and whether that's something that _should_ be handled. 

I'm assuming the first option is the one we'd probably care about here.

**Comment by nerrad** [comment]
> we should specify which Payment is crashing, we have that data; the message should tell the merchant to contact those plugin maintainers

If possible, yes. I agree with Thomas though that any notice should only be visible to a viewer that can action on that (not shoppers).

> the notice should not be displayed in the button listing. but above it. The result is not ideal. I see that the ErrorBoundry is at <li> level so the PR will need some logic to move that error a level up

I agree, but keep in mind that's where it was before the changes in the PR (https://github.com/woocommerce/woocommerce/pull/42988/files#diff-67dc69fb379e1605451feb74dafb0f094bbb2d666e39873517f75faf580200e0L164-L168). As I understand it this might require using something other than an ErrorBoundary since the intention of the PR is to allow payment methods that are working okay to display. What guidance would you give in that case Raluca?

> this extra logic would enable us to also not display the Express Payments container in case no payment button works, which by the way should be handled in this PR as well as we need to cover all cases

That's what the changes in this PR already enable. I think the extra guidance (as I suggest previously) would be helpful here since it's _likely_ an ErrorBoundary won't accomplish the requirements. I have some thoughts on what the logic could be but would be interested in your take on that first.

**Comment by elizaan36** [comment]
hey, +1 to @ralucaStan 's [comment here](https://github.com/woocommerce/woocommerce/pull/42988#issuecomment-1971498535). I think the error message should be merchant-facing only and indicate which payment method has the issue.

I'm not super clear on messaging but in terms of placement of the error, could we consider showing it in the sidebar instead? There's a precedent with plugin incompatibilities. Is there a possibility that multiple payment methods could have the same error?

image

**Comment by tarunvijwani** [comment]
Thank you all for your valuable feedback and suggestions. Based on the discussion and the recommendations provided, I'm closing this PR and creating a new one with a fresh approach that incorporates the improvements discussed.

**Pull request closed without merging**
